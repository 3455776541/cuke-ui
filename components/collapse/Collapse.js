import React from "react";
import cls from "classnames";
import PropTypes from "prop-types";

export default class Collapse extends React.PureComponent {
  state = {
    activeKey: [],
    currentActiveKey: ""
  };
  static defaultProps = {
    prefixCls: "cuke-collapse",
    defaultActiveKey: [],
    activeKey: [],
    disabled: false,
    hideArrow: false,
    accordion: false
  };
  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    defaultActiveKey: PropTypes.array,
    activeKey: PropTypes.array,
    onChange: PropTypes.func,
    disabled: PropTypes.bool,
    hideArrow: PropTypes.bool,
    accordion: PropTypes.bool
  };

  onChange = key => {
    let activeKey = [...this.state.activeKey];
    if (!activeKey.includes(key)) {
      activeKey.push(key);
    } else {
      activeKey = activeKey.filter(_activeKey => _activeKey !== key);
    }

    this.setState({ activeKey, currentActiveKey: key });

    if (this.props.onChange) {
      this.props.onChange(activeKey);
    }
  };
  render() {
    const {
      className,
      prefixCls,
      children,
      accordion,
      defaultActiveKey,
      activeKey,
      disabled,
      hideArrow,
      ...attr
    } = this.props;

    const { currentActiveKey } = this.state;

    const items = React.Children.map(children, (element, index) => {
      return React.cloneElement(element, {
        key: index,
        accordion,
        activeKey: String(index + 1),
        disabled: element.props.disabled || disabled,
        hideArrow: element.props.hideArrow || hideArrow,
        visible: accordion
          ? currentActiveKey === String(index + 1)
          : defaultActiveKey.includes(String(index + 1)) ||
          activeKey.includes(String(index + 1)), //eslint-disable-line
        onChange: this.onChange
      });
    });

    return (
      <div
        className={cls(prefixCls, className, {
          [`${prefixCls}-accordion`]: accordion
        })}
        {...attr}
      >
        {items}
      </div>
    );
  }
}
