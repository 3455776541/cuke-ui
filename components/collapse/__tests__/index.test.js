import React from "react";
import { render, shallow, mount } from "enzyme";
import assert from "power-assert";
import toJson from "enzyme-to-json";
import Collapse from "../index";

describe("<Collapse/>", () => {
  it("should render Collapse", () => {
    const wrapper = render(
      <Collapse>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-collapse class name", () => {
    const wrapper = shallow(
      <Collapse>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    assert(wrapper.find(".cuke-collapse").length === 1);
  });

  it("should render Collapse.Item", () => {
    const wrapper = shallow(
      <Collapse>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    assert(wrapper.find(Collapse.Item).length === 2);
  });

  it("should render default active key", () => {
    const wrapper = mount(
      <Collapse defaultActiveKey={["1", "2"]}>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    assert(wrapper.find(".cuke-collapse-item-arrow-active").length === 2);
  });

  it("should render active key", () => {
    const wrapper = mount(
      <Collapse defaultActiveKey={["1"]}>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    assert(wrapper.find(".cuke-collapse-item-arrow-active").length === 1);
  });

  it("should render hide arrow for collapseItem", () => {
    const wrapper = mount(
      <Collapse>
        <Collapse.Item title="标题1" hideArrow>
          内容1
				</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    assert(wrapper.find(".cuke-collapse-item-arrow").length === 1);
  });

  it("should render hide arrow for collapse", () => {
    const wrapper = mount(
      <Collapse hideArrow>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    assert(wrapper.find(".cuke-collapse-item-arrow").length === 0);
  });

  it("should disabled for collapse", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Collapse disabled onChange={onChange}>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    wrapper.find(".cuke-collapse").simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });

  it("should accordion mode", () => {
    const wrapper = shallow(
      <Collapse accordion>
        <Collapse.Item title="标题1">内容1</Collapse.Item>
        <Collapse.Item title="标题2">内容2</Collapse.Item>
      </Collapse>
    );
    wrapper.find(".cuke-collapse").simulate("click");
    assert(wrapper.find(".cuke-collapse-accordion").length === 1);
  });
});
