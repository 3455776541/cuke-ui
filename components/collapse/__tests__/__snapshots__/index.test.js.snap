// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse/> should render Collapse 1`] = `
<div
  class="cuke-collapse"
>
  <div
    class="cuke-collapse-item"
  >
    <div
      class="cuke-collapse-item-header cuke-collapse-item-header-has-radius"
    >
      <span
        class="cuke-collapse-item-arrow"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </span>
      标题1
    </div>
    <div
      class="cuke-collapse-item-content cuke-collapse-item-hide"
    >
      内容1
    </div>
  </div>
  <div
    class="cuke-collapse-item"
  >
    <div
      class="cuke-collapse-item-header cuke-collapse-item-header-has-radius"
    >
      <span
        class="cuke-collapse-item-arrow"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </span>
      标题2
    </div>
    <div
      class="cuke-collapse-item-content cuke-collapse-item-hide"
    >
      内容2
    </div>
  </div>
</div>
`;
